### PR TITLE
[SPARK-42791][SQL] Create a new golden file test framework for analysis

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR updates the `SQLQueryTestSuite` to also consume the same input SQL queries from the input files in second pass and then perform analysis and generate the string representation of the analyzed plans, in a separate new directory. It works much the same way as the previous `SQLQueryTestSuite` behavior except that it now also produces analyzed plan string instead of just query results in the output golden files.

### Why are the changes needed?

This framework will help us guard against bugs in future development by showing a clear signal when analyzer updates result in changes to the query plans for a body of test queries. PR authors and reviewers will be able to see the diffs for all changed plans during the review, and any unintentional plan changes will act as a signal for PR authors to adjust their code to prevent the changes from happening.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This test adds a new test suite and initial golden file. The new test suite passes as of this PR.

### CODE REVIEW

**Comment by dtenedor** [comment]
Hi @gengliangwang this should be ready for a first look!

**Comment by gengliangwang** [comment]
@dtenedor Since we already have `SQLQueryTestSuite` which has good basic Spark SQL features coverage, shall we combine both? E.g. let `SQLQueryTestSuite` show analyzed plan/optimized plan/execution results for comparison, and port more tests from [zetasql](https://github.com/google/zetasql/tree/master/zetasql/analyzer/testdata)
What is the reason for having a new analysis test only?

**Comment by dtenedor** [comment]
@gengliangwang Sure, I was thinking about this too. We can reuse the same input SQL query files if we want, and just generate and test against different analyzer test output files. Let me update the PR to do that and I can ping the thread again.

**Comment by dtenedor** [comment]
@gengliangwang from past experience we will want to keep the query plans separate from the SQL results, otherwise the SQL results become hard to read. I will put the analyzer results in separate files.

**Comment by gengliangwang** [comment]
> I will put the analyzer results in separate files.

Sounds great! Thanks for the work!

**Comment by dtenedor** [comment]
@gengliangwang alright I made this change, please look again when you are ready.

File: sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala [hunk 1]
```diff
@@ -228,6 +229,101 @@ abstract class QueryTest extends PlanTest {
     assert(query.queryExecution.executedPlan.missingInput.isEmpty,
       s"The physical plan has missing inputs:\n${query.queryExecution.executedPlan}")
   }
+
+  /** A single SQL query's output. */
+  protected trait QueryTestOutput {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala:234]: Now all the changes in `QueryTest` is only used in `SQLQueryTestSuite`. Shall we move these changes to `SQLQueryTestSuite`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 2]
```diff
@@ -205,6 +194,14 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
   protected case class RegularTestCase(
       name: String, inputFile: String, resultFile: String) extends TestCase
 
+  /** An ANSI-related test case. */
+  protected case class AnsiTestCase(
+      name: String, inputFile: String, resultFile: String) extends TestCase with AnsiTest
+
+  /** An analyzer test that shows the analyzed plan string as output. */
+  protected case class AnalyzerTestCase(
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:202]: hmm, will we have analyzer-only test? Or you will remove this one when generating all the analyzer golden files?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 3]
```diff
@@ -525,23 +403,18 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
         AnsiTestCase(testCaseName, absPath, resultFile) :: Nil
       } else if (file.getAbsolutePath.startsWith(s"$inputFilePath${File.separator}timestampNTZ")) {
         TimestampNTZTestCase(testCaseName, absPath, resultFile) :: Nil
+      } else if (analyzerTestCaseList.contains(file.getName.toLowerCase(Locale.ROOT))) {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:406]: I think analyzerTestCaseList should cover the `ansi`/`TimestampNTZ`/`postgreSQL` input queries as well

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 2 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:202]: In this PR I made the test runner generate both a `RegularTestCase` and one of these new `AnalyzerTestCase` instances for each of the eligible input `.sql` files.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 3 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:406]: Good idea, done!

**Review by dtenedor — COMMENTED** [review]
Thanks @gengliangwang for your review!

File: sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala [hunk 4]
```diff
@@ -228,6 +228,7 @@ abstract class QueryTest extends PlanTest {
     assert(query.queryExecution.executedPlan.missingInput.isEmpty,
       s"The physical plan has missing inputs:\n${query.queryExecution.executedPlan}")
   }
+
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala:231]: Let remove the unnecessary change

File: sql/core/src/test/resources/sql-tests/analyzer-results/array.sql.out
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/analyzer-results/array.sql.out:1]: I expect there is `ansi/array.sql.out` as well

Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/analyzer-results/array.sql.out:1]: Yes, just uploaded this.

**Review by gengliangwang — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Also, cc @sunchao

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala [hunk 5]
```diff
@@ -108,4 +126,23 @@ trait SQLQueryTestHelper {
...
+  }
+
+  protected def splitCommentsAndCodes(input: String): (Array[String], Array[String]) =
+    input.split("\n").partition { line =>
+      val newLine = line.trim
+      newLine.startsWith("--") && !newLine.startsWith("--QUERY-DELIMITER")
+    }
+
+  /** Returns all the files (not directories) in a directory, recursively. */
+  protected def listFilesRecursively(path: File): Seq[File] = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:141]: I assume these 3 method are just copied from the existing code

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 6]
```diff
@@ -678,4 +663,99 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
...
+
+  /** A single SQL query's output. */
+  trait QueryTestOutput {
+    def sql: String
+    def schema: Option[String]
+    def output: String
+    def numSegments: Int
+  }
+
+  /** A single SQL query's output. */
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:730]:
```suggestion
  /** A single SQL query's execution output. */
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 7]
```diff
@@ -266,22 +261,27 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
...
-    def splitWithSemicolon(seq: Seq[String]) = {
-      seq.mkString("\n").split("(?<=[^\\\\]);")
-    }
-
-    def splitCommentsAndCodes(input: String) = input.split("\n").partition { line =>
-      val newLine = line.trim
-      newLine.startsWith("--") && !newLine.startsWith("--QUERY-DELIMITER")
-    }
+  /** A test case. */
+  protected trait TestCase {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:270]: can we move it back to the original place? https://github.com/apache/spark/pull/40449/files#diff-212b35a074b1eeb518aff4c47130c7ab9cc23856ea6ea62f083af4452c637b9aL178

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 8]
```diff
@@ -338,6 +338,7 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
       conf.trim -> value.substring(1).trim
     })
 
+    val regenerateGoldenFiles: Boolean = System.getenv("SPARK_GENERATE_GOLDEN_FILES") == "1"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:341]: why do we redefine the variable here?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 9]
```diff
@@ -266,22 +261,27 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
...
+    val name: String
+    val inputFile: String
+    val resultFile: String
+  }
 
+  /** Run a test case. */
+  protected def runSqlTestCase(
+      testCase: TestCase,
+      listTestCases: Seq[TestCase],
+      runQueries: (
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:280]: why do we pass it as a parameter? We can directly invoke `runQueries` in `runSqlTestCase`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 10]
```diff
@@ -460,49 +473,25 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
...
     listFilesRecursively(new File(inputFilePath)).flatMap { file =>
       val resultFile = file.getAbsolutePath.replace(inputFilePath, goldenFilePath) + ".out"
+      val analyzerResultFile =
+        file.getAbsolutePath.replace(inputFilePath, analyzerGoldenFilePath) + ".out"
       val absPath = file.getAbsolutePath
       val testCaseName = absPath.stripPrefix(inputFilePath).stripPrefix(File.separator)
+      val analyzerTestCaseName = s"${testCaseName}_analyzer_test"
 
+      val newTestCase =
       if (file.getAbsolutePath.startsWith(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:495]:
```suggestion
      val newTestCase = if (file.getAbsolutePath.startsWith(
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 11]
```diff
@@ -678,4 +663,99 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
       super.afterAll()
     }
   }
+
+  /**
+   * Consumes contents from a single golden file and compares the expected results against the
+   * output of running a query.
+   */
+  def readGoldenFileAndCompareResults(
+    resultFile: String,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:672]: nit: 4 spaces indentation for parameters.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala (hunk 5 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:141]: Yes, I had moved these methods from their original place elsewhere in this file. I moved them back now to simplify the diff.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 9 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:280]: Good point, I removed this and just call `runQueries` directly now.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 11 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:672]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 6 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:730]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 10 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:495]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 8 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:341]: removed this.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 7 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:270]: Done.

**Review by dtenedor — COMMENTED** [review]
Thanks @cloud-fan for your review! Addressed comments.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by LuciferYang** [comment]
https://github.com/apache/spark/pull/40492

**Comment by dtenedor** [comment]
Looks like some extra tests got added just as this was getting merged! Thanks @LuciferYang for this fix 👍
